File: shared/api/library.ts
import { invoke } from '@tauri-apps/api/core'
import type {
	DuplicateResolution,
	FileMatchResult,
	ImportResult,
	ImportResultWithDuplicates,
	LibraryFolderScanResult,
	Track,
	TrackColor,
	TrackFilter,
	TrackUpdate,
} from '../types'

/**
 * Import tracks from file paths into the library
 */
export async function importTracks(paths: string[]): Promise<ImportResult> {
	return invoke<ImportResult>('import_tracks', { paths })
}

/**
 * Get all tracks with optional filtering
 */
export async function getTracks(filter?: TrackFilter): Promise<Track[]> {
	return invoke<Track[]>('get_tracks', { filter: filter ?? null })
}

/**
 * Get a single track by ID
 */
export async function getTrack(id: string): Promise<Track> {
	return invoke<Track>('get_track', { id })
}

/**
 * Update track metadata
 */
export async function updateTrack(id: string, update: TrackUpdate): Promise<Track> {
	return invoke<Track>('update_track', { id, update })
}

/**
 * Delete tracks by IDs
 */
export async function deleteTracks(ids: string[]): Promise<void> {
	return invoke<void>('delete_tracks', { ids })
}

/**
 * Search tracks by query string
 */
export async function searchTracks(query: string): Promise<Track[]> {
	return invoke<Track[]>('search_tracks', { query })
}

/**
 * Result of a rescan artwork operation
 */
export interface RescanResult {
	updated_count: number
	failed_count: number
}

/**
 * Rescan artwork for all tracks that don't have artwork yet
 */
export async function rescanArtwork(): Promise<RescanResult> {
	return invoke<RescanResult>('rescan_artwork')
}

/**
 * Rescan artwork for a single track
 */
export async function rescanTrackArtwork(id: string): Promise<boolean> {
	return invoke<boolean>('rescan_track_artwork', { id })
}

/**
 * Check if a track's file exists on disk
 */
export async function checkFileExists(trackId: string): Promise<boolean> {
	return invoke<boolean>('check_file_exists', { trackId })
}

/**
 * Validate if a replacement file matches the original track
 */
export async function validateReplacementFile(trackId: string, newPath: string): Promise<FileMatchResult> {
	return invoke<FileMatchResult>('validate_replacement_file', { trackId, newPath })
}

/**
 * Relocate a track to a new file path
 */
export async function relocateTrack(trackId: string, newPath: string, force: boolean = false): Promise<Track> {
	return invoke<Track>('relocate_track', { trackId, newPath, force })
}

/**
 * Set color for multiple tracks
 * @param trackIds - Array of track IDs to update
 * @param color - Color to set (null to remove color)
 */
export async function setTrackColors(trackIds: string[], color: TrackColor | null): Promise<void> {
	return invoke<void>('set_track_colors', { trackIds, color })
}

/**
 * Update multiple tracks with the same update data (bulk operation)
 */
export async function updateTracks(ids: string[], update: TrackUpdate): Promise<Track[]> {
	return invoke<Track[]>('update_tracks', { ids, update })
}

/**
 * Set artwork for a track from a user-provided file
 */
export async function setTrackArtwork(trackId: string, filePath: string): Promise<Track> {
	return invoke<Track>('set_track_artwork', { trackId, filePath })
}

/**
 * Delete artwork for a track
 */
export async function deleteTrackArtwork(trackId: string): Promise<Track> {
	return invoke<Track>('delete_track_artwork', { trackId })
}

/**
 * Re-extract artwork from the audio file
 */
export async function reextractTrackArtwork(trackId: string): Promise<Track> {
	return invoke<Track>('reextract_track_artwork', { trackId })
}

/**
 * Compare artwork files for multiple tracks to check if they are identical.
 * Returns the shared artwork path if all tracks have identical artwork, or null otherwise.
 */
export async function compareTrackArtworks(trackIds: string[]): Promise<string | null> {
	return invoke<string | null>('compare_track_artworks', { trackIds })
}

/**
 * Import tracks with duplicate detection based on content hash
 */
export async function importTracksWithDuplicates(paths: string[]): Promise<ImportResultWithDuplicates> {
	return invoke<ImportResultWithDuplicates>('import_tracks_with_duplicates', { paths })
}

/**
 * Resolve a duplicate track with the user's chosen action
 */
export async function resolveDuplicate(resolution: DuplicateResolution): Promise<Track | null> {
	return invoke<Track | null>('resolve_duplicate', { resolution })
}

/**
 * Get the music folder configured for this device, if any
 */
export async function getMusicLibraryFolder(): Promise<string | null> {
	return invoke<string | null>('get_music_library_folder')
}

/**
 * Persist this device's music folder and return the canonical path actually stored
 */
export async function setMusicLibraryFolder(path: string): Promise<string> {
	return invoke<string>('set_music_library_folder', { path })
}

/**
 * Recursively import supported audio files from the configured music folder.
 * Files already in the library are skipped silently by the backend.
 */
export async function scanMusicLibraryFolder(): Promise<LibraryFolderScanResult> {
	return invoke<LibraryFolderScanResult>('scan_music_library_folder')
}

File: src-tauri/src/commands/library.rs
use std::path::PathBuf;

use tauri::State;

use crate::error::Result;
use crate::models::{
    DuplicateResolution, FileMatchResult, ImportResult, ImportResultWithDuplicates, Track,
    TrackFilter, TrackUpdate,
};
use crate::services::library::{LibraryFolderScanResult, RescanResult};
use crate::services::LibraryService;

#[tauri::command]
pub async fn import_tracks(
    paths: Vec<String>,
    library: State<'_, LibraryService>,
) -> Result<ImportResult> {
    let pathbufs: Vec<PathBuf> = paths.into_iter().map(PathBuf::from).collect();
    library.import_tracks(pathbufs)
}

#[tauri::command]
pub async fn get_tracks(
    filter: Option<TrackFilter>,
    library: State<'_, LibraryService>,
) -> Result<Vec<Track>> {
    library.get_tracks(filter)
}

#[tauri::command]
pub async fn get_track(id: String, library: State<'_, LibraryService>) -> Result<Track> {
    library.get_track(&id)
}

#[tauri::command]
pub async fn update_track(
    id: String,
    update: TrackUpdate,
    library: State<'_, LibraryService>,
) -> Result<Track> {
    library.update_track(&id, update)
}

#[tauri::command]
pub async fn delete_tracks(ids: Vec<String>, library: State<'_, LibraryService>) -> Result<()> {
    library.delete_tracks(ids)
}

#[tauri::command]
pub async fn search_tracks(
    query: String,
    library: State<'_, LibraryService>,
) -> Result<Vec<Track>> {
    let filter = TrackFilter {
        search: Some(query),
        ..Default::default()
    };
    library.get_tracks(Some(filter))
}

#[tauri::command]
pub async fn rescan_artwork(library: State<'_, LibraryService>) -> Result<RescanResult> {
    library.rescan_all_artwork()
}

#[tauri::command]
pub async fn rescan_track_artwork(id: String, library: State<'_, LibraryService>) -> Result<bool> {
    library.rescan_track_artwork(&id)
}

#[tauri::command]
pub async fn check_file_exists(
    track_id: String,
    library: State<'_, LibraryService>,
) -> Result<bool> {
    library.check_track_file_exists(&track_id)
}

#[tauri::command]
pub async fn validate_replacement_file(
    track_id: String,
    new_path: String,
    library: State<'_, LibraryService>,
) -> Result<FileMatchResult> {
    library.validate_replacement_file(&track_id, &PathBuf::from(new_path))
}

#[tauri::command]
pub async fn relocate_track(
    track_id: String,
    new_path: String,
    force: bool,
    library: State<'_, LibraryService>,
) -> Result<Track> {
    library.relocate_track(&track_id, &PathBuf::from(new_path), force)
}

#[tauri::command]
pub async fn set_track_colors(
    track_ids: Vec<String>,
    color: Option<String>,
    library: State<'_, LibraryService>,
) -> Result<()> {
    library.set_track_colors(track_ids, color)
}

#[tauri::command]
pub async fn update_tracks(
    ids: Vec<String>,
    update: TrackUpdate,
    library: State<'_, LibraryService>,
) -> Result<Vec<Track>> {
    library.update_tracks(ids, update)
}

#[tauri::command]
pub async fn set_track_artwork(
    track_id: String,
    file_path: String,
    library: State<'_, LibraryService>,
) -> Result<Track> {
    library.set_track_artwork(&track_id, &PathBuf::from(file_path))
}

#[tauri::command]
pub async fn delete_track_artwork(
    track_id: String,
    library: State<'_, LibraryService>,
) -> Result<Track> {
    library.delete_track_artwork(&track_id)
}

#[tauri::command]
pub async fn reextract_track_artwork(
    track_id: String,
    library: State<'_, LibraryService>,
) -> Result<Track> {
    library.reextract_track_artwork(&track_id)
}

#[tauri::command]
pub async fn compare_track_artworks(
    track_ids: Vec<String>,
    library: State<'_, LibraryService>,
) -> Result<Option<String>> {
    library.compare_track_artworks(&track_ids)
}

#[tauri::command]
pub async fn import_tracks_with_duplicates(
    paths: Vec<String>,
    library: State<'_, LibraryService>,
) -> Result<ImportResultWithDuplicates> {
    let pathbufs: Vec<PathBuf> = paths.into_iter().map(PathBuf::from).collect();
    library.import_tracks_with_duplicate_detection(pathbufs)
}

#[tauri::command]
pub async fn resolve_duplicate(
    resolution: DuplicateResolution,
    library: State<'_, LibraryService>,
) -> Result<Option<Track>> {
    library.resolve_duplicate(resolution)
}

#[tauri::command]
pub async fn get_music_library_folder(
    library: State<'_, LibraryService>,
) -> Result<Option<String>> {
    library.music_library_folder()
}

#[tauri::command]
pub async fn set_music_library_folder(
    path: String,
    library: State<'_, LibraryService>,
) -> Result<String> {
    library.set_music_library_folder(&path)
}

#[tauri::command]
pub async fn scan_music_library_folder(
    library: State<'_, LibraryService>,
) -> Result<LibraryFolderScanResult> {
    library.scan_music_library_folder()
}
